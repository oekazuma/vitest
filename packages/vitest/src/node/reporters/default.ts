import type { UserConsoleLog } from '../../types'
import { BaseReporter } from './base'
import type { ListRendererOptions } from './renderers/listRenderer'
import { createListRenderer } from './renderers/listRenderer'

export class DefaultReporter extends BaseReporter {
  renderer?: ReturnType<typeof createListRenderer>
  rendererOptions: ListRendererOptions = {} as any

  onCollected() {
    if (this.isTTY) {
      this.rendererOptions.outputStream = this.ctx.outputStream
      const files = this.ctx.state.getFiles(this.watchFilters)
      if (!this.renderer)
        this.renderer = createListRenderer(files, this.rendererOptions).start()
      else
        this.renderer.update(files)
    }
  }

  async onFinished(files = this.ctx.state.getFiles()) {
    await this.stopListRender()
    this.ctx.log()
    await super.onFinished(files)
  }

  async onWatcherStart() {
    await this.stopListRender()
    super.onWatcherStart()
  }

  async stopListRender() {
    await this.renderer?.stop()
    this.renderer = undefined
  }

  async onWatcherRerun(files: string[], trigger?: string) {
    await this.stopListRender()
    await super.onWatcherRerun(files, trigger)
  }

  onUserConsoleLog(log: UserConsoleLog) {
    this.renderer?.clear()
    super.onUserConsoleLog(log)
  }
}
