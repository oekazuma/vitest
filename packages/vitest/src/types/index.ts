export * from './config'
export * from './tasks'
export * from './reporter'
export * from './snapshot'
export * from './worker'
export * from './general'
export * from './coverage'
