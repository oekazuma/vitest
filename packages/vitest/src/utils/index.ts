import c from 'picocolors'
import { isPackageExists } from 'local-pkg'
import { resolve } from 'pathe'
import type { Suite, Task } from '../types'
import { getNames } from './tasks'

export * from './tasks'
export * from './path'
export * from './base'

export const isWindows = process.platform === 'win32'

/**
 * Partition in tasks groups by consecutive concurrent
 */
export function partitionSuiteChildren(suite: Suite) {
  let tasksGroup: Task[] = []
  const tasksGroups: Task[][] = []
  for (const c of suite.tasks) {
    if (tasksGroup.length === 0 || c.concurrent === tasksGroup[0].concurrent) {
      tasksGroup.push(c)
    }
    else {
      tasksGroups.push(tasksGroup)
      tasksGroup = [c]
    }
  }
  if (tasksGroup.length > 0)
    tasksGroups.push(tasksGroup)

  return tasksGroups
}

export function getFullName(task: Task) {
  return getNames(task).join(c.dim(' > '))
}

export async function ensurePackageInstalled(
  dependency: string,
  promptInstall = !process.env.CI && process.stdout.isTTY,
) {
  if (isPackageExists(dependency))
    return true

  process.stderr.write(c.red(`${c.inverse(c.red(' MISSING DEP '))} Can not find dependency '${dependency}'\n\n`))

  if (!promptInstall)
    return false

  const prompts = await import('prompts')
  const { install } = await prompts.prompt({
    type: 'confirm',
    name: 'install',
    message: c.reset(`Do you want to install ${c.green(dependency)}?`),
  })

  if (install) {
    await (await import('@antfu/install-pkg')).installPackage(dependency, { dev: true })
    // TODO: somehow it fails to load the package after installation, remove this when it's fixed
    process.stderr.write(c.yellow(`\nPackage ${dependency} installed, re-run the command to start.\n`))
    process.exit(1)
    return true
  }

  return false
}

/**
 * If code starts with a function call, will return its last index, respecting arguments.
 * This will return 25 - last ending character of toMatch ")"
 * Also works with callbacks
 * ```
 * toMatch({ test: '123' });
 * toBeAliased('123')
 * ```
 */
export function getCallLastIndex(code: string) {
  let charIndex = -1
  let inString: string | null = null
  let startedBracers = 0
  let endedBracers = 0
  let beforeChar: string | null = null
  while (charIndex <= code.length) {
    beforeChar = code[charIndex]
    charIndex++
    const char = code[charIndex]

    const isCharString = char === '"' || char === '\'' || char === '`'

    if (isCharString && beforeChar !== '\\') {
      if (inString === char)
        inString = null
      else if (!inString)
        inString = char
    }

    if (!inString) {
      if (char === '(')
        startedBracers++
      if (char === ')')
        endedBracers++
    }

    if (startedBracers && endedBracers && startedBracers === endedBracers)
      return charIndex
  }
  return null
}

export { resolve as resolvePath }
