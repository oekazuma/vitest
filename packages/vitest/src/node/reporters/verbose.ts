import { DefaultReporter } from './default'

export class VerboseReporter extends DefaultReporter {
  constructor() {
    super()
    this.rendererOptions.renderSucceed = true
  }
}
