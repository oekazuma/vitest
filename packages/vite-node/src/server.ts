import type { TransformResult, ViteDevServer } from 'vite'
import type { FetchResult, RawSourceMap, ViteNodeResolveId, ViteNodeServerOptions } from './types'
import { shouldExternalize } from './externalize'
import { toFilePath, withInlineSourcemap } from './utils'

export * from './externalize'

export class ViteNodeServer {
  private fetchPromiseMap = new Map<string, Promise<FetchResult>>()
  private transformPromiseMap = new Map<string, Promise<TransformResult | null | undefined>>()

  fetchCache = new Map<string, {
    timestamp: number
    result: FetchResult
  }>()

  constructor(
    public server: ViteDevServer,
    public options: ViteNodeServerOptions = {},
  ) {}

  shouldExternalize(id: string) {
    return shouldExternalize(id, this.options.deps)
  }

  async resolveId(id: string, importer?: string): Promise<ViteNodeResolveId | null> {
    return this.server.pluginContainer.resolveId(id, importer, { ssr: true })
  }

  async fetchModule(id: string): Promise<FetchResult> {
    // reuse transform for concurrent requests
    if (!this.fetchPromiseMap.has(id)) {
      this.fetchPromiseMap.set(id,
        this._fetchModule(id)
          .then((r) => {
            return this.options.sourcemap !== true ? { ...r, map: undefined } : r
          })
          .finally(() => {
            this.fetchPromiseMap.delete(id)
          }),
      )
    }
    return this.fetchPromiseMap.get(id)!
  }

  async transformRequest(id: string) {
    // reuse transform for concurrent requests
    if (!this.transformPromiseMap.has(id)) {
      this.transformPromiseMap.set(id,
        this._transformRequest(id)
          .finally(() => {
            this.transformPromiseMap.delete(id)
          }),
      )
    }
    return this.transformPromiseMap.get(id)!
  }

  getTransformMode(id: string) {
    const withoutQuery = id.split('?')[0]

    if (this.options.transformMode?.web?.some(r => withoutQuery.match(r)))
      return 'web'
    if (this.options.transformMode?.ssr?.some(r => withoutQuery.match(r)))
      return 'ssr'

    if (withoutQuery.match(/\.([cm]?[jt]sx?|json)$/))
      return 'ssr'
    return 'web'
  }

  private async _fetchModule(id: string): Promise<FetchResult> {
    let result: FetchResult

    const filePath = toFilePath(id, this.server.config.root)

    const module = this.server.moduleGraph.getModuleById(id)
    const timestamp = module?.lastHMRTimestamp || Date.now()
    const cache = this.fetchCache.get(filePath)
    if (timestamp && cache && cache.timestamp >= timestamp)
      return cache.result

    const externalize = await this.shouldExternalize(filePath)
    if (externalize) {
      result = { externalize }
    }
    else {
      const r = await this._transformRequest(id)
      result = { code: r?.code, map: r?.map as unknown as RawSourceMap }
    }

    this.fetchCache.set(filePath, {
      timestamp,
      result,
    })

    return result
  }

  private async _transformRequest(id: string) {
    let result: TransformResult | null = null

    if (this.getTransformMode(id) === 'web') {
      // for components like Vue, we want to use the client side
      // plugins but then convert the code to be consumed by the server
      result = await this.server.transformRequest(id)
      if (result)
        result = await this.server.ssrTransform(result.code, result.map, id)
    }
    else {
      result = await this.server.transformRequest(id, { ssr: true })
    }

    const sourcemap = this.options.sourcemap ?? 'inline'
    if (sourcemap === 'inline' && result && !id.includes('node_modules'))
      withInlineSourcemap(result)

    return result
  }
}
