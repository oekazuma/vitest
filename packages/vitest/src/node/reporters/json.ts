import { promises as fs } from 'fs'
import { resolve } from 'pathe'
import type { Vitest } from '../../node'
import type { File, Reporter } from '../../types'
import { getSuites, getTests } from '../../utils'

// for compatibility reasons, the reporter produces a JSON similar to the one produced by the Jest JSON reporter
// the following types are extracted from the Jest repository (and simplified)
interface TestResult {
  displayName?: string
  failureMessage?: string | null
  skipped: boolean
  status?: string
  testFilePath?: string
}
interface AggregatedResult {
  numFailedTests: number
  numFailedTestSuites: number
  numPassedTests: number
  numPassedTestSuites: number
  numPendingTests: number
  numTodoTests: number
  numPendingTestSuites: number
  numTotalTests: number
  numTotalTestSuites: number
  startTime: number
  success: boolean
  testResults: Array<TestResult>
}

export class JsonReporter implements Reporter {
  start = 0
  ctx!: Vitest

  onInit(ctx: Vitest): void {
    this.ctx = ctx
    this.start = Date.now()
  }

  protected async logTasks(files: File[]) {
    const suites = getSuites(files)
    const numTotalTestSuites = suites.length
    const tests = getTests(files)
    const numTotalTests = tests.length

    const numFailedTestSuites = suites.filter(s => s.result?.error).length
    const numPassedTestSuites = numTotalTestSuites - numFailedTestSuites
    const numPendingTestSuites = suites.filter(s => s.result?.state === 'run').length
    const numFailedTests = tests.filter(t => t.result?.state === 'fail').length
    const numPassedTests = numTotalTests - numFailedTests
    const numPendingTests = tests.filter(t => t.result?.state === 'run').length
    const numTodoTests = tests.filter(t => t.mode === 'todo').length

    const success = numFailedTestSuites === 0 && numFailedTests === 0

    const testResults: Array<TestResult> = tests.map(t => ({
      displayName: t.name,
      failureMessage: t.result?.error?.message,
      skipped: t.result?.state === 'skip',
      status: t.result?.state,
      testFilePath: t.file?.filepath,
    }))

    const result: AggregatedResult = { numTotalTestSuites, numPassedTestSuites, numFailedTestSuites, numPendingTestSuites, numTotalTests, numPassedTests, numFailedTests, numPendingTests, numTodoTests, startTime: this.start, success, testResults }

    await this.writeReport(JSON.stringify(result, null, 2))
  }

  async onFinished(files = this.ctx.state.getFiles()) {
    await this.logTasks(files)
  }

  /**
   * Writes the report to an output file if specified in the config,
   * or logs it to the console otherwise.
   * @param report
   */
  async writeReport(report: string) {
    if (this.ctx.config.outputFile) {
      const reportFile = resolve(this.ctx.config.root, this.ctx.config.outputFile)
      await fs.writeFile(reportFile, report, 'utf-8')
      this.ctx.log(`JSON report written to ${reportFile}`)
    }
    else {
      this.ctx.log(report)
    }
  }
}
