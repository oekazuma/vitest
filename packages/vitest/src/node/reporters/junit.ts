import { promises as fs } from 'fs'
import { hostname } from 'os'
import { relative, resolve } from 'pathe'

import type { Vitest } from '../../node'
import type { ErrorWithDiff, Reporter, Task } from '../../types'
import { parseStacktrace } from '../../utils/source-map'
import { F_POINTER } from '../../utils/figures'
import { IndentedLogger } from './renderers/indented-logger'

function flattenTasks(task: Task, baseName = ''): Task[] {
  const base = baseName ? `${baseName} > ` : ''

  if (task.type === 'suite') {
    return task.tasks.flatMap(child => flattenTasks(child, `${base}${task.name}`))
  }
  else {
    return [{
      ...task,
      name: `${base}${task.name}`,
    }]
  }
}

function escapeXML(value: any): string {
  return String(value)
    .replace(/&/g, '&amp;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
}

function getDuration(task: Task): string | undefined {
  return task.result?.duration ? (task.result.duration / 1000).toFixed(10) : undefined
}

export class JUnitReporter implements Reporter {
  private ctx!: Vitest
  private reportFile?: string
  private baseLog!: (text: string) => Promise<void>
  private logger!: IndentedLogger<Promise<void>>

  async onInit(ctx: Vitest): Promise<void> {
    this.ctx = ctx

    if (this.ctx.config.outputFile) {
      this.reportFile = resolve(this.ctx.config.root, this.ctx.config.outputFile)
      const fileFd = await fs.open(this.reportFile, 'w+')

      this.baseLog = async(text: string) => await fs.writeFile(fileFd, `${text}\n`)
    }
    else {
      this.baseLog = async(text: string) => this.ctx.log(text)
    }

    this.logger = new IndentedLogger(this.baseLog)
  }

  async writeElement(name: string, attrs: Record<string, any>, children: () => Promise<void>) {
    const pairs = []
    for (const key in attrs) {
      const attr = attrs[key]
      if (attr === undefined)
        continue

      pairs.push(`${key}="${escapeXML(attr)}"`)
    }

    await this.logger.log(`<${name}${pairs.length ? ` ${pairs.join(' ')}` : ''}>`)
    this.logger.indent()
    await children.call(this)
    this.logger.unindent()

    await this.logger.log(`</${name}>`)
  }

  async writeErrorDetails(error: ErrorWithDiff): Promise<void> {
    const errorName = error.name ?? error.nameStr ?? 'Unknown Error'
    await this.baseLog(`${errorName}: ${error.message}`)

    const stack = parseStacktrace(error)

    // TODO: This is same as printStack but without colors. Find a way to reuse code.
    for (const frame of stack) {
      const pos = frame.sourcePos ?? frame
      const path = relative(this.ctx.config.root, frame.file)

      await this.baseLog(` ${F_POINTER} ${[frame.method, `${path}:${pos.line}:${pos.column}`].filter(Boolean).join(' ')}`)

      // reached at test file, skip the follow stack
      if (frame.file in this.ctx.state.filesMap)
        break
    }
  }

  async writeLogs(task: Task, type: 'err' | 'out'): Promise<void> {
    if (task.logs == null || task.logs.length === 0)
      return

    const logType = type === 'err' ? 'stderr' : 'stdout'
    const logs = task.logs.filter(log => log.type === logType)

    if (logs.length === 0)
      return

    await this.writeElement(`system-${type}`, {}, async() => {
      for (const log of logs)
        await this.baseLog(escapeXML(log.content))
    })
  }

  async writeTasks(tasks: Task[], filename: string): Promise<void> {
    for (const task of tasks) {
      await this.writeElement('testcase', {
        classname: filename,
        name: task.name,
        time: getDuration(task),
      }, async() => {
        await this.writeLogs(task, 'out')
        await this.writeLogs(task, 'err')

        if (task.mode === 'skip' || task.mode === 'todo')
          this.logger.log('<skipped/>')

        if (task.result?.state === 'fail') {
          const error = task.result.error

          await this.writeElement('failure', {
            message: error?.message,
            type: error?.name ?? error?.nameStr,
          }, async() => {
            if (!error)
              return

            await this.writeErrorDetails(error)
          })
        }
      })
    }
  }

  async onFinished(files = this.ctx.state.getFiles()) {
    await this.logger.log('<?xml version="1.0" encoding="UTF-8" ?>')

    const transformed = files
      .map((file) => {
        const tasks = file.tasks.flatMap(task => flattenTasks(task))

        const stats = tasks.reduce((stats, task) => {
          return {
            passed: stats.passed + Number(task.result?.state === 'pass'),
            failures: stats.failures + Number(task.result?.state === 'fail'),
            skipped: stats.skipped + Number(task.mode === 'skip' || task.mode === 'todo'),
          }
        },
        {
          passed: 0,
          failures: 0,
          skipped: 0,
        })

        return {
          ...file,
          tasks,
          stats,
        }
      })

    await this.writeElement('testsuites', {}, async() => {
      for (const file of transformed) {
        await this.writeElement('testsuite', {
          name: file.name,
          timestamp: (new Date()).toISOString(),
          hostname: hostname(),
          tests: file.tasks.length,
          failures: file.stats.failures,
          errors: 0, // An errored test is one that had an unanticipated problem. We cannot detect those.
          skipped: file.stats.skipped,
          time: getDuration(file),
        }, async() => {
          await this.writeTasks(file.tasks, file.name)
        })
      }
    })

    if (this.reportFile)
      this.ctx.log(`JUNIT report written to ${this.reportFile}`)
  }
}
