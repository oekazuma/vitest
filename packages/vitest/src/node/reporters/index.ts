import { DefaultReporter } from './default'
import { DotReporter } from './dot'
import { JsonReporter } from './json'
import { VerboseReporter } from './verbose'
import { TapReporter } from './tap'
import { JUnitReporter } from './junit'
import { TapFlatReporter } from './tap-flat'

export { DefaultReporter }

export const ReportersMap = {
  'default': DefaultReporter,
  'verbose': VerboseReporter,
  'dot': DotReporter,
  'json': JsonReporter,
  'tap': TapReporter,
  'tap-flat': TapFlatReporter,
  'junit': JUnitReporter,
}

export type BuiltinReporters = keyof typeof ReportersMap
