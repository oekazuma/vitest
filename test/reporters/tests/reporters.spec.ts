import { afterEach, expect, test, vi } from 'vitest'
import { JsonReporter } from '../../../packages/vitest/src/node/reporters/json'
import { JUnitReporter } from '../../../packages/vitest/src/node/reporters/junit'
import { TapReporter } from '../../../packages/vitest/src/node/reporters/tap'
import { TapFlatReporter } from '../../../packages/vitest/src/node/reporters/tap-flat'
import { getContext } from '../src/context'
import { files } from '../src/data'

afterEach(() => {
  vi.useRealTimers()
})

test('tap reporter', async() => {
  // Arrange
  const reporter = new TapReporter()
  const context = getContext()

  // Act
  reporter.onInit(context.vitest)
  await reporter.onFinished(files)

  // Assert
  expect(context.output).toMatchSnapshot()
})

test('tap-flat reporter', async() => {
  // Arrange
  const reporter = new TapFlatReporter()
  const context = getContext()

  // Act
  reporter.onInit(context.vitest)
  await reporter.onFinished(files)

  // Assert
  expect(context.output).toMatchSnapshot()
})

test('JUnit reporter', async() => {
  // Arrange
  const reporter = new JUnitReporter()
  const context = getContext()

  vi.mock('os', () => ({
    hostname: () => 'hostname',
  }))

  vi.setSystemTime(1642587001759)

  // Act
  reporter.onInit(context.vitest)
  await reporter.onFinished(files)

  // Assert
  expect(context.output).toMatchSnapshot()
})

test('json reporter', async() => {
  // Arrange
  const reporter = new JsonReporter()
  const context = getContext()

  vi.setSystemTime(1642587001759)

  // Act
  reporter.onInit(context.vitest)
  await reporter.onFinished(files)

  // Assert
  expect(context.output).toMatchSnapshot()
})
