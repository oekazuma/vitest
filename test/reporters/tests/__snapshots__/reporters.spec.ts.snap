// Vitest Snapshot v1

exports[`JUnit reporter 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" ?>
<testsuites>
    <testsuite name=\\"test/core/test/basic.test.ts\\" timestamp=\\"2022-01-19T10:10:01.759Z\\" hostname=\\"hostname\\" tests=\\"8\\" failures=\\"1\\" errors=\\"0\\" skipped=\\"1\\" time=\\"0.1459928420\\">
        <testcase classname=\\"test/core/test/basic.test.ts\\" name=\\"suite &gt; inner suite &gt; Math.sqrt()\\" time=\\"0.0014422860\\">
            <failure message=\\"expected 2.23606797749979 to equal 2\\" type=\\"AssertionError\\">
AssertionError: expected 2.23606797749979 to equal 2
 ‚ùØ vitest/test/core/test/basic.test.ts:8:32
            </failure>
        </testcase>
        <testcase classname=\\"test/core/test/basic.test.ts\\" name=\\"suite &gt; JSON\\" time=\\"0.0010237110\\">
        </testcase>
        <testcase classname=\\"test/core/test/basic.test.ts\\" name=\\"suite &gt; async with timeout\\">
            <skipped/>
        </testcase>
        <testcase classname=\\"test/core/test/basic.test.ts\\" name=\\"suite &gt; timeout\\" time=\\"0.1005059841\\">
        </testcase>
        <testcase classname=\\"test/core/test/basic.test.ts\\" name=\\"suite &gt; callback setup success \\" time=\\"0.0201848750\\">
        </testcase>
        <testcase classname=\\"test/core/test/basic.test.ts\\" name=\\"suite &gt; callback test success \\" time=\\"0.0003324542\\">
        </testcase>
        <testcase classname=\\"test/core/test/basic.test.ts\\" name=\\"suite &gt; callback setup success done(false)\\" time=\\"0.0197386060\\">
        </testcase>
        <testcase classname=\\"test/core/test/basic.test.ts\\" name=\\"suite &gt; callback test success done(false)\\" time=\\"0.0001923509\\">
        </testcase>
    </testsuite>
</testsuites>
"
`;

exports[`json reporter 1`] = `
"{
  \\"numTotalTestSuites\\": 3,
  \\"numPassedTestSuites\\": 3,
  \\"numFailedTestSuites\\": 0,
  \\"numPendingTestSuites\\": 0,
  \\"numTotalTests\\": 8,
  \\"numPassedTests\\": 7,
  \\"numFailedTests\\": 1,
  \\"numPendingTests\\": 0,
  \\"numTodoTests\\": 0,
  \\"startTime\\": 1642587001759,
  \\"success\\": false,
  \\"testResults\\": [
    {
      \\"displayName\\": \\"Math.sqrt()\\",
      \\"failureMessage\\": \\"expected 2.23606797749979 to equal 2\\",
      \\"skipped\\": false,
      \\"status\\": \\"fail\\",
      \\"testFilePath\\": \\"/vitest/test/core/test/basic.test.ts\\"
    },
    {
      \\"displayName\\": \\"JSON\\",
      \\"skipped\\": false,
      \\"status\\": \\"pass\\",
      \\"testFilePath\\": \\"/vitest/test/core/test/basic.test.ts\\"
    },
    {
      \\"displayName\\": \\"async with timeout\\",
      \\"skipped\\": false,
      \\"testFilePath\\": \\"/vitest/test/core/test/basic.test.ts\\"
    },
    {
      \\"displayName\\": \\"timeout\\",
      \\"skipped\\": false,
      \\"status\\": \\"pass\\",
      \\"testFilePath\\": \\"/vitest/test/core/test/basic.test.ts\\"
    },
    {
      \\"displayName\\": \\"callback setup success \\",
      \\"skipped\\": false,
      \\"status\\": \\"pass\\",
      \\"testFilePath\\": \\"/vitest/test/core/test/basic.test.ts\\"
    },
    {
      \\"displayName\\": \\"callback test success \\",
      \\"skipped\\": false,
      \\"status\\": \\"pass\\",
      \\"testFilePath\\": \\"/vitest/test/core/test/basic.test.ts\\"
    },
    {
      \\"displayName\\": \\"callback setup success done(false)\\",
      \\"skipped\\": false,
      \\"status\\": \\"pass\\",
      \\"testFilePath\\": \\"/vitest/test/core/test/basic.test.ts\\"
    },
    {
      \\"displayName\\": \\"callback test success done(false)\\",
      \\"skipped\\": false,
      \\"status\\": \\"pass\\",
      \\"testFilePath\\": \\"/vitest/test/core/test/basic.test.ts\\"
    }
  ]
}
"
`;

exports[`tap reporter 1`] = `
"TAP version 13
1..1
not ok 1 - test/core/test/basic.test.ts # time=145.99ms {
    1..1
    ok 1 - suite # time=1.90ms {
        1..8
        ok 1 - inner suite # time=1.90ms {
            1..1
            not ok 1 - Math.sqrt() # time=1.44ms
                ---
                error:
                    name: \\"AssertionError\\"
                    message: \\"expected 2.23606797749979 to equal 2\\"
                at: \\"/vitest/test/core/test/basic.test.ts:8:32\\"
                actual: \\"2.23606797749979\\"
                expected: \\"2\\"
                ...
        }
        ok 2 - JSON # time=1.02ms
        ok 3 - async with timeout # SKIP
        ok 4 - timeout # time=100.51ms
        ok 5 - callback setup success  # time=20.18ms
        ok 6 - callback test success  # time=0.33ms
        ok 7 - callback setup success done(false) # time=19.74ms
        ok 8 - callback test success done(false) # time=0.19ms
    }
}
"
`;

exports[`tap-flat reporter 1`] = `
"TAP version 13
1..8
not ok 1 - test/core/test/basic.test.ts > suite > inner suite > Math.sqrt() # time=1.44ms
    ---
    error:
        name: \\"AssertionError\\"
        message: \\"expected 2.23606797749979 to equal 2\\"
    at: \\"/vitest/test/core/test/basic.test.ts:8:32\\"
    actual: \\"2.23606797749979\\"
    expected: \\"2\\"
    ...
ok 2 - test/core/test/basic.test.ts > suite > JSON # time=1.02ms
ok 3 - test/core/test/basic.test.ts > suite > async with timeout # SKIP
ok 4 - test/core/test/basic.test.ts > suite > timeout # time=100.51ms
ok 5 - test/core/test/basic.test.ts > suite > callback setup success  # time=20.18ms
ok 6 - test/core/test/basic.test.ts > suite > callback test success  # time=0.33ms
ok 7 - test/core/test/basic.test.ts > suite > callback setup success done(false) # time=19.74ms
ok 8 - test/core/test/basic.test.ts > suite > callback test success done(false) # time=0.19ms
"
`;
